### PR TITLE
qemu-vm: remove bootDisk, refactor using make-disk-image with better EFI support

### PR DESCRIPTION
**Pull request by RaitoBezarius**
### Description of changes

This PR does multiple things :

###### make-disk-image refactoring in QEMU test infrastructure

As per https://github.com/NixOS/nixpkgs/issues/23052 — QEMU test infrastructure do not use ad-hoc code to build the boot disk image but reuse make-disk-image which was improved to support use-cases.

Doing this consolidates make-disk-image as one place to do these changes, it seems like the boot disk production code was doing old stuff with GRUB 0.97, I'm uncertain whether that should be cleaned up.

###### UEFI firmware infrastructure

As part of https://github.com/NixOS/nixpkgs/pull/187887 ; this uses consolidated UEFI firmware and variables and expose their variables to end users.

It also adds SecureBoot and System Management Mode enforcement support in the test infrastructure and the make-disk-image function.

EFIVARS can be modified as part of the disk image build process, enabling people to build a disk image with specific authenticated UEFI variables, e.g. SecureBoot keys.

###### Documentation

This PR adds documentation on what are the breaking changes in regard to the disk layout (storeImage is used only if needed at all.)

TODO/notes:

- [x] Cleanup UEFI firmware infrastructure: done in
- [x] Cleanup the /dev/vdb2 stuff in QEMU code.
- [x] Some tests such as installer depends on bootDevice being vdb or second drive.
- [x] GRUB version 1 may have some subtleties.
- [x] Assert touchEFIVars only if partition table type is hybrid or efi.
- [x] Use a better error message for assert.
- [x] Refactor in make-disk-image the EFI firmware infrastructure using https://github.com/NixOS/nixpkgs/pull/187887.
- [x] Simple tests:
- [x] Legacy boot without any EFI stuff should still work.
- [x] EFI/Hybrid boot without touching EFI vars should still work.
- [x] EFI/Hybrid boot with touching EFI vars should still work AND we should be able to write/read in EFIVARS.
- [x] Reproducibility of system image, wrt to fixed GUIDs in partition table.
- [x] Documentation:
    - [x] How to test this kind of changes? What to look for?
    - [x] Where are the breaking changes? Everything that relies on /dev/vdb2 is now incorrect.
    - [x] Compatibility layer? Deprecation notice?
    - [x] Prevent users to shoot their foot: increase the number of assertions & warnings accordingly in the module
- [ ] For a future PR
    - [ ] Over {systemd-boot, GRUB1, GRUB2} cartesian product for SecureBoot test.
    - [ ] makeEc2Test is useful to initialize using an image, it would be good to extract it as a lower level primitive to start test off an already existing disk image, maybe put it directly into makeTest machinery or makeTestFromImage

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Opened due to GitHub merging the old PR #191665

**Comment by RaitoBezarius** [comment]
With this, https://hydra.nixos.org/eval/1786610 ; this PR do not introduce any failing test beyond what there was on the base nixpkgs point of this PR.

Will fix the merge conflicts now.

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 1]
```diff
@@ -198,103 +212,35 @@ let
 
   regInfo = pkgs.closureInfo { rootPaths = config.virtualisation.additionalPaths; };
 
-
   # Generate a hard disk image containing a /boot partition and GRUB
   # in the MBR.  Used when the `useBootLoader' option is set.
   # Uses `runInLinuxVM` to create the image in a throwaway VM.
   # See note [Disk layout with `useBootLoader`].
```
Comment (nikstur) [inline nixos/modules/virtualisation/qemu-vm.nix:218]: This is not really accurate anymore. The Disk Layout note is gone and the image might not contain grub, right?

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 2]
```diff
@@ -111,8 +115,22 @@ let
...
+          echo "Disk image do not exist, creating the virtualisation disk image..."
+          # If we are using a bootloader and default filesystems layout.
+          # We have to reuse the system image layout as a backing image format (CoW)
+          # And we can write on the top of it, this way.
+          # Otherwise, if we are not using the default filesystem layout, maybe, we want the raw disk
+          # and we do boot using the system image direct kernels (useBootLoader == false),
+          # but we want to perform operations in the postDeviceCommands
+          # to format the raw disk, etc. So we will not depend on the system image layout as a backing image format.
+          # Also, if we are not using a bootloader, we can use the size parameter as we are not using CoW anymore.
+          # CoW prevent size to be affected to an image, TODO: is this a bug?
```
Comment (nikstur) [inline nixos/modules/virtualisation/qemu-vm.nix:127]: This comment could use some cleanup so that its easier to read.

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 3]
```diff
@@ -198,103 +212,35 @@ let
...
+    partitionTableType = selectPartitionTableLayout { inherit (cfg) useDefaultFilesystems useEFIBoot; };
+    installBootLoader = cfg.useBootLoader && cfg.useDefaultFilesystems;
+    touchEFIVars = cfg.useEFIBoot;
+    diskSize = "auto";
+    additionalSpace = "0M";
+    copyChannel = false;
+    OVMF = cfg.efi.OVMF;
+  };
 
+  # This image is supposed to have no side-effects on the boot disk.
```
Comment (nikstur) [inline nixos/modules/virtualisation/qemu-vm.nix:234]:
```suggestion
  # This image is supposed to have no side-effects on the system image.
```

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 4]
```diff
@@ -979,23 +994,14 @@ in
         file = ''"$NIX_DISK_IMAGE"'';
         driveExtraOpts.cache = "writeback";
         driveExtraOpts.werror = "report";
+        deviceExtraOpts.bootindex = mkIf cfg.useBootLoader "1";
```
Comment (nikstur) [inline nixos/modules/virtualisation/qemu-vm.nix:997]: Shouldn't this always be bootindex=1 ?

**Review by nikstur — APPROVED** [review]
This is a very good change and I'm happy to see it. Most of my comments are just minor nitpicks.

File: nixos/modules/virtualisation/qemu-vm.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix:218]: Right!

File: nixos/modules/virtualisation/qemu-vm.nix (hunk 4 above)
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix:997]: I have no a good mental model of how bootindex works in interaction with direct kernel boot. :/
Will investigate.

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 5]
```diff
@@ -305,7 +307,9 @@ In addition to numerous new and upgraded packages, this release includes the fol
 
 - The `bloat` package has been updated from unstable-2022-03-31 to unstable-2022-10-25, which brings a breaking change. See [this upstream commit message](https://git.freesoftwareextremist.com/bloat/commit/?id=887ed241d64ba5db3fd3d87194fb5595e5ad7d73) for details.
 
-- Synapse's systemd unit has been hardened.
+- `pkgs.OVMF.fd` exposes `firmware` and `variables` which points to your host architecture binaries for the correspoding UEFI artifacts.
+
+- The `services.matrix-synapse` systemd unit has been hardened.
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2211.section.md:312]: Huh, I should remove new things I introduced here.

File: nixos/lib/make-disk-image.nix [hunk 6]
```diff
@@ -70,6 +82,13 @@
 , # Disk image format, one of qcow2, qcow2-compressed, vdi, vpc, raw.
   format ? "raw"
 
+  # Whether to fix partitions GUIDs/UUIDs to arbitrary values.
```
Comment (samueldr) [inline nixos/lib/make-disk-image.nix:88]: (More general task than for this line)

Be mindful of the confusing confusion that can happen with:

 - GPT Partition **Unique Identifier** [in the form of a UUID/GUID]
 - GPT Partition **Type** identifier [in the form for a GUID]
 - Filesystem **Unique Identifier** [filesystem dependent]

Can you ensure **every** instance of wording around those is properly identified to not be confused with one eachother? For example, never use the short form "partition guid" or worse, guid or uuid; but always *partition unique identifier*, *partition type identifier* and *filesystem unique identifier*?

Not necessarily for your own sake, but for the sake of everyone else re-reading this.

File: nixos/lib/make-disk-image.nix [hunk 7]
```diff
@@ -70,6 +82,13 @@
...
+  # Also, to fix last time checked of the ext4 partition if fsType = ext4.
+, deterministic ? true
+
+  # GUID used for root partition
```
Comment (samueldr) [inline nixos/lib/make-disk-image.nix:92]: Is this a *Partition Unique Identifier* for use with `deterministic`?

**Comment by samueldr** [comment]
One overall note, still reading this whopper of a PR: I think the lack of separation of concerns by shoving *filesystem creation duties* with *disk image creation duties* in the same expression/derivation is likely to cause issues overall. And we're seeing even less separation of concern by addition of irrelevant *machine state* within this, I don't really like this.

File: nixos/lib/make-disk-image.nix (hunk 6 above)
Comment (RaitoBezarius) [reply nixos/lib/make-disk-image.nix:88]: You are totally right!

**Comment by RaitoBezarius** [comment]
> One overall note, still reading this whopper of a PR: I think the lack of separation of concerns by shoving _filesystem creation duties_ with _disk image creation duties_ in the same expression/derivation is likely to cause issues overall. And we're seeing even less separation of concern by addition of irrelevant _machine state_ (EFI vars) within this, I don't really like this.

I agree with you, but this PR has only a small ambition (which took already a lot of time to do "well"), that is, to remove the ad-hoc disk image creation that existed before and put the knobs into that super function (which is bad, I know.)

Once it lands, we can:

- split disk image creation *and* filesystem creation (*with machine state*), to expose a lower level function
- split filesystem creation and generic machine state recording during VM operations

IMHO, it's easier to deal with it in this way also because I have a use case (SecureBoot) which can nourish the adequate refactors, some of them were already mentioned in the N previous PRs (that were closed due to my mistakes), see also https://github.com/NixOS/nixpkgs/pull/187855 which is an interesting symptom exposing ad-hoc knobs to test *images*, but no generic tooling to do this across the ecosystem.

**Comment by samueldr** [comment]
I believe it would be better if this fully owns the fact it's adding more to the technical debt, with the goal to make the tests we need better. And then that instead of breaking out chunks *from this implementation*, we use the time this buys us to properly investigate options forward for disk image generation.

Note that I'm not saying this while doing nothing. Within three projects of mine, I have already started stumbling and hitting my face against either bad starts or thorny problems.

**Comment by RaitoBezarius** [comment]
> I believe it would be better if this fully owns the fact it's adding more to the technical debt, with the goal to make the tests we need better. And then that instead of breaking out chunks _from this implementation_, we use the time this buys us to properly investigate options forward for disk image generation.
> 
> Note that I'm not saying this while doing nothing. Within three projects of mine, I have already started stumbling and hitting my face against either bad starts or thorny problems.

Fair, I see your point. 

Nevertheless, do you think it's futile spending time on making these images more deterministic?

**Comment by samueldr** [comment]
> Nevertheless, do you think it's futile spending time on making these images more deterministic?

Futile no. I think it's good to want to improve what is there.

I think my main thought is that *this is the wrong model*.

As a cleanup effort, a short-to-mid term effort, great! But building too large on top of this foundation is (in my opinion) not the way forward.

**Comment by roberth** [comment]
> the lack of separation of concerns by shoving _filesystem creation duties_ with _disk image creation duties_ in the same expression/derivation is likely to cause issues overall.

There's a technical argument to be made for not producing a lot of garbage in the store. Such images fill up the store quickly, requiring more frequent garbage collection, reducing the utility of the local store as a cache. This gets even worse for CIs, which push build dependencies, including such images, to a persistent cache.

**Comment by samueldr** [comment]
> There's a technical argument to be made for not producing a lot of garbage in the store. Such images fill up the store quickly, requiring more frequent garbage collection, reducing the utility of the local store as a cache. This gets even worse for CIs, which push build dependencies, including such images, to a persistent cache.

You're right. I have been slowly working on a scheme to make the *concerns* separate, but coalesce in a single derivation. So this *should* be a non-issue in the end.

In other words, the user interface separates concerns, the implementation builds on discrete blocks, but they all (can) assemble in a single derivation.

This is *anyway* a requirement for it to properly work with the org's hydra and the ARM sd image, as the discrete components are too big to be cached. And as you said, it causes a lot of unneeded churn.

File: nixos/lib/make-disk-image.nix (hunk 6 above)
Comment (RaitoBezarius) [reply nixos/lib/make-disk-image.nix:88]: Done.

File: nixos/lib/make-disk-image.nix (hunk 7 above)
Comment (RaitoBezarius) [reply nixos/lib/make-disk-image.nix]: Clarified.

**Comment by RaitoBezarius** [comment]
Ready for another round of review.

File: nixos/lib/make-disk-image.nix [hunk 8]
```diff
@@ -17,7 +17,7 @@
   # either "efi" or "hybrid"
   # This will be undersized slightly, as this is actually the offset of
   # the end of the partition. Generally it will be 1MiB smaller.
-  bootSize ? "256MiB"
+  bootSize ? "256M"
```
Comment (Mic92) [inline nixos/lib/make-disk-image.nix:20]: In https://github.com/NixOS/nixpkgs/pull/203641/commits/2900974e7604b57ff332e2401f7f32cde8303611 you are not just changing documentation but also revert `

```diff
- bootSize ? "256MiB"
+ bootSize ? "256M"
```

This should be in https://github.com/NixOS/nixpkgs/pull/203641/commits/5380f26a8b806b3ac80769b3e3094812e76e97bb

File: nixos/tests/secureboot.nix [hunk 9]
```diff
@@ -16,7 +16,7 @@ let
     };
     boot.loader.systemd-boot.enable = true;
     boot.loader.efi.canTouchEfiVariables = true;
-    environment.systemPackages = [ pkgs.efibootmgr pkgs.sbsigntool pkgs.sbctl ];
+    environment.systemPackages = with pkgs; [ efibootmgr sbsigntool sbctl ];
```
Comment (Mic92) [inline nixos/tests/secureboot.nix:19]: This also should not be in a commit that changes documentation but in your secureboot test commit.

File: nixos/maintainers/scripts/ec2/amazon-image.nix [hunk 10]
```diff
@@ -43,7 +43,7 @@ in {
 
     sizeMB = mkOption {
       type = with types; either (enum [ "auto" ]) int;
-      default = 2048;
+      default = 2252;
```
Comment (Mic92) [inline nixos/maintainers/scripts/ec2/amazon-image.nix:46]: The commit does not say why the disk size was bumped.

File: doc/builders/images.xml [hunk 11]
```diff
@@ -10,4 +10,5 @@
  <xi:include href="images/ocitools.section.xml" />
  <xi:include href="images/snaptools.section.xml" />
  <xi:include href="images/portableservice.section.xml" />
+ <xi:include href="images/makediskimage.section.xml" />
```
Comment (Mic92) [inline doc/builders/images.xml:13]: Did you change anything about the interface? Otherwise I would pull this out into its own PR.

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 12]
```diff
@@ -316,14 +316,34 @@ in
           '';
         };
 
+    virtualisation.bootPartition =
```
Comment (Mic92) [inline nixos/modules/virtualisation/qemu-vm.nix:319]: Was the legacy boot broken before or broken by your changes?

File: doc/builders/images.xml (hunk 11 above)
Comment (RaitoBezarius) [reply doc/builders/images.xml:13]: Yes, I added parameters.

File: nixos/maintainers/scripts/ec2/amazon-image.nix (hunk 10 above)
Comment (RaitoBezarius) [reply nixos/maintainers/scripts/ec2/amazon-image.nix:46]: Indeed, the Amazon test was broken due to not enough disk space, I don't know if it was induced by me or in-between latest release and master. Will double check.

File: nixos/modules/virtualisation/qemu-vm.nix (hunk 12 above)
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix:319]: Broken by my changes

File: nixos/tests/secureboot.nix (hunk 9 above)
Comment (RaitoBezarius) [reply nixos/tests/secureboot.nix:19]: True

File: nixos/lib/make-disk-image.nix (hunk 8 above)
Comment (RaitoBezarius) [reply nixos/lib/make-disk-image.nix:20]: Right

**Comment by RaitoBezarius** [comment]
Closed in favor of #207039 #207038 #207043.
